### DISCUSSION TITLE
WIND table entries in grib2 file

### DISCUSSION THREAD

**Discussion by gsketefian** [dtcenter/METplus#2813] [File I/O] [answered]
Hi,

I'm using METplus to run GridStat on a set of NetCDF forecast files and a set of grib2 "obs" files that are in fact from the HRRR.  I am having trouble understanding how MET determines what obs field to use for upper-level horizontal wind speed (WIND).

In my GridStat configuration file, the parameters for this WIND variable are set as follows:
```
FCST_VAR4_NAME = WIND
FCST_VAR4_LEVELS = "(0,11,*,*)", "(0,10,*,*)", "(0,9,*,*)", "(0,8,*,*)", "(0,7,*,*)", "(0,6,*,*)", "(0,5,*,*)", "(0,4,*,*)", "(0,3,*,*)", "(0,2,*,*)", "(0,1,*,*)", "(0,0,*,*)"
FCST_VAR4_THRESH = ge2.572, ge2.572&&lt5.144, ge5.144, ge10.288, ge15.433, ge20.577, ge25.722
FCST_VAR4_OPTIONS = set_attr_lead = "{lead?fmt=%H%M%S}";

OBS_VAR4_NAME = WIND
OBS_VAR4_LEVELS = P1000, P975, P950, P925, P900, P875, P850, P825, P800, P475, P300, P200
OBS_VAR4_THRESH = ge2.572, ge2.572&&lt5.144, ge5.144, ge10.288, ge15.433, ge20.577, ge25.722
```
Running this gives me the following message in the METplus log file regarding the HRRR "obs" file:
``` 
DEBUG 3: Multiple GRIB2 table entries match lookup criteria (parm_name = WIND):
DEBUG 3:   parm_name: WIND, index_a = 0, grib2_mtab = 1, grib2_cntr = 0, grib2_ltab = 0, index_b = 2, index_c = 1
DEBUG 3:   parm_name: WIND, index_a = 0, grib2_mtab = 14, grib2_cntr = 57, grib2_ltab = 0, index_b = 2, index_c = 1
DEBUG 3:   parm_name: WIND, index_a = 0, grib2_mtab = 1, grib2_cntr = 7, grib2_ltab = 1, index_b = 2, index_c = 1
DEBUG 3: Using the first match found:   parm_name: WIND, index_a = 0, grib2_mtab = 1, grib2_cntr = 0, grib2_ltab = 0, index_b = 2, index_c = 1
```
So it seems like METplus is finding 3 different GRIB2 table entries for WIND in the grib2 file, but I am not sure how those entries map onto the one WIND field in the grib2 file, which is:
```
$ wgrib2 -v2 /scratch2/BMC/fv3lam/Gerard.Ketefian/using_hrrr_as_obs/hrrr/20240401/21/hrrr.t21z.wrfprsf01.grib2 | grep WIND   
624:332249478:21Z01apr2024:WIND Wind Speed [m/s]:lvl1=(103,10) lvl2=(255,missing):10 m above ground:0-1 hour max fcst:
```
This wind field is at 10m above ground, while in my conf file settings I specify various pressure levels but not L10.  So I am wondering what those 3 table entries are that METplus is finding.  I'm happy to provide more information about the contents of the grib2, etc.

Thanks.
@JeffBeck-NOAA @VanderleiVargas-NOAA

**Comment by JohnHalleyGotway** [comment, marked as answer]
@gsketefian I see you have a question about the GRIB table lookup logic. And I'll note that your question very likely relates to this existing GitHub issue dtcenter/MET#2240. We did start work on this issue a while back, but never straightened out the details with enough confidence to get it across the finish line... knowing that changing the existing GRIB table lookup logic would introduce significant risk in "breaking" existing usage of the tools.

Thanks for sending along some details and output, but I'm not quite convinced yet that MET is comparing the "wrong" fields yet.

I ran the following commands on hera...

Setup environment:
```
module load intel/2022.1.2
module use -a /contrib/met/modulefiles/
module load met/11.1.1
module load wgrib2
```
Run `plot_data_plane` to read 10-m WIND speed:
```
plot_data_plane /scratch2/BMC/fv3lam/Gerard.Ketefian/using_hrrr_as_obs/hrrr/20240401/21/hrrr.t21z.wrfprsf01.grib2 wind_Z10.ps 'name="WIND"; level="Z10";' -v 3
```
And note the following log message indicating that it's directly reading record number 624 of the input GRIB2 file:
```
DEBUG 3: MetGrib2DataFile::data_plane() -> Found exact match for 'WIND/Z10' in GRIB2 record 624 field 1 of GRIB2 file '/scratch2/BMC/fv3lam/Gerard.Ketefian/using_hrrr_as_obs/hrrr/20240401/21/hrrr.t21z.wrfprsf01.grib2'
```
That is consistent with the output from wgrib2 for 10-m winds.
Next, run `plot_data_plane` to read WIND speed for 1000mb, which does NOT appear in the output from wgrib2:
```
plot_data_plane /scratch2/BMC/fv3lam/Gerard.Ketefian/using_hrrr_as_obs/hrrr/20240401/21/hrrr.t21z.wrfprsf01.grib2 wind_P1000.ps 'name="WIND"; level="P1000";' -v 3
```
And note the following log messages indicating that MET is automatically deriving WIND speed from the corresponding U/V components, after first rotating those components from grid-to-earth relative:
```
DEBUG 3: MetGrib2DataFile::data_plane_array() -> Found exact match for "WIND/P1000" in GRIB2 record 538 field 1 of GRIB2 file "/scratch2/BMC/fv3lam/Gerard.Ketefian/using_hrrr_as_obs/hrrr/20240401/21/hrrr.t21z.wrfprsf01.grib2"
DEBUG 3: MetGrib2DataFile::check_uv_rotation() -> Found pair match "VGRD/P1000" in GRIB2 record 539 field 1
DEBUG 3: Rotating U and V wind components from grid-relative to earth-relative.
DEBUG 3: MetGrib2DataFile::data_plane_array() -> Found exact match for "WIND/P1000" in GRIB2 record 539 field 1 of GRIB2 file "/scratch2/BMC/fv3lam/Gerard.Ketefian/using_hrrr_as_obs/hrrr/20240401/21/hrrr.t21z.wrfprsf01.grib2"
DEBUG 3: MetGrib2DataFile::check_uv_rotation() -> Found pair match "UGRD/P1000" in GRIB2 record 538 field 1
DEBUG 3: Rotating U and V wind components from grid-relative to earth-relative.
DEBUG 3: Deriving wind speed from U and V wind components.
DEBUG 3: MetGrib2DataFile::data_plane() -> Found derived match for 'WIND/P1000' in GRIB2 file '/scratch2/BMC/fv3lam/Gerard.Ketefian/using_hrrr_as_obs/hrrr/20240401/21/hrrr.t21z.wrfprsf01.grib2'
```

So lots of things going on here!

When reading wind speed (`WIND`) or wind direction (`WDIR`) from GRIB1 or GRIB2 files, the MET library code uses the following logic:
1. Search the input GRIB file for an exact match, and if present, use it. This is occurring for 10-m wind in your input file.
2. If not present, search for the corresponding components named `UGRD` and `VGRD`. And when reading those, rotate them from grid-to-earth relative, if needed. And then derive the requested wind field. This is occurring for the pressure level winds in your input file.

Let me adds a few notes:
- I agree that the extra log messages about GRIB table lookup details are misleading, and I really would like to refine that logic via dtcenter/MET#2240. But in this case, they're a red herring. Grid-Stat really is using the data you want it to use. Run it at verbosity level 3 or higher and inspect the log messages to confirm.
- The auto-derivation of wind speed and direction only currently works well for GRIB1/2 inputs with U and V components named `UGRD` and `VGRD`. Those settings are hard-coded. I'd really like to enhance this functionality in several ways, as described in these [MET: Wind Verification](https://github.com/dtcenter/MET/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+label%3A%22MET%3A+Wind+Verification%22) issues. Big picture, this derivation functionality should be extended to support other input file formats (like NetCDF), alternative GRIB conventions (like `U` and `V` in ECMWF files), and all input data type variations (like wind speed, direction, U-wind, or V-wind) present in gridded or point forecasts or observations. If any of these a priorities for your DTC-related work, we could line up DTC project funding to support that work.
- The logic for handling winds is not described in the MET User's Guide at a sufficient level of detail. The work for the issues linked above should include updating the documentation to clearly describe how winds are handled in MET.

Hopefully that helps clarify.

**Reply by JohnHalleyGotway** [reply]
@gsketefian, just following up on this discussion. Do additional questions remain? Or can I mark this as being answered and close the discussion?

**Comment by gsketefian** [comment]
Hi John,

Thanks for the detailed answer, and apologies for not getting back to you.
Yes, you can mark this as answered and close the discussion.

Gerard


On Mon, Jan 13, 2025 at 2:29 PM John Halley Gotway ***@***.***>
wrote:

> @gsketefian <https://github.com/gsketefian>, just following up on this
> discussion. Do additional questions remain? Or can I mark this as being
> answered and close the discussion?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/2813#discussioncomment-11824951>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHM3ZYQOHDKXI6UAL4EERDT2KQV2PAVCNFSM6AAAAABTDTPXG6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCOBSGQ4TKMI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Gerard Ketefian
Research Scientist
NOAA/OAR/ESRL/GSD/EMB, R/GSD1
325 Broadway
Boulder, CO 80305
phone: 970-703-3048
